aiutils/tftools/batch_normalizer.py

```diff
@@ -95,3 +97,62 @@ def batch_normalize(input, phase_train, name):
     return control_flow_ops.cond(phase_train,
                                  lambda: bn.normalize(input, True),
                                  lambda: bn.normalize(input, False))
+
+
+class BatchNorm():
+    def __init__(self, input, training, decay=0.95, epsilon=1e-4, name='bn'):
+        self.decay = decay
+        self.epsilon = epsilon
+        self.batchnorm(input, training, name)
+
+    def batchnorm(self, input, training, name):
+        with tf.name_scope(name) as bn:
+            rank = len(input.get_shape().as_list())
+            in_dim = input.get_shape().as_list()[-1]
+
+            if rank == 2:
+                print('rank2')
+                axes = [0]
+            elif rank == 4:
+                print('rank4')
+                axes = [0, 1, 2]
+            else:
+                raise ValueError('Input tensor must have rank 2 or 4.')
+
+            self.mean, self.variance = tf.nn.moments(input, axes)
+            self.offset = tf.Variable(initial_value=tf.constant(
+                value=0.0, shape=[in_dim]),
+                                      name='offset')
+            self.scale = tf.Variable(initial_value=tf.constant(value=1.0,
+                                                               shape=[in_dim]),
+                                     name='scale')
+
+            self.ema = tf.train.ExponentialMovingAverage(decay=self.decay)
+
+            self.output = tf.cond(training,
+                                  lambda: self.get_normalizer(input, True),
+                                  lambda: self.get_normalizer(input, False))
+
+    def get_normalizer(self, input, train_flag):
+        if train_flag:
+            ema_apply_op = self.ema.apply([self.mean, self.variance])
+            with tf.control_dependencies([ema_apply_op]):
+                self.output_training = tf.nn.batch_normalization(
+                    input, self.mean, self.variance, self.offset, self.scale,
```

Comment:
How would you feel about moving the definition of self.mean and self.variance inside this, and then moving them to local scope instead of object scope?

Then the only thing at object scope would be offset, scale, ema, and output which I think makes more sense.